lasagne/layers/recurrent.py

```diff
@@ -0,0 +1,1239 @@
+            raise ValueError("Input sequence length cannot be specified as "
+                             "None when unroll_scan is True")
+
+        # Check that output shapes match
+        if input_to_hidden.output_shape != hidden_to_hidden.output_shape:
+            raise ValueError("The output shape for input_to_hidden and "
+                             "input_to_hidden must be equal, but "
+                             "input_to_hidden.output_shape={} and "
+                             "hidden_to_hidden.output_shape={}".format(
+                                 input_to_hidden.output_shape,
+                                 hidden_to_hidden.output_shape))
+
+        if nonlinearity is None:
+            self.nonlinearity = nonlinearities.identity
+        else:
+            self.nonlinearity = nonlinearity
+
+        # Get the input dimensionality and number of units based on the
+        # expected output of the input-to-hidden layer
+        self.num_inputs = np.prod(self.input_shape[2:])
+        self.num_units = input_to_hidden.output_shape[-1]
+
+        # Initialize hidden state
+        if isinstance(hid_init, T.TensorVariable):
+            if hid_init.ndim != 2:
+                raise ValueError(
+                    "When hid_init is provided as a TensorVariable, it should "
+                    "have 2 dimensions and have shape (num_batch, num_units)")
+            self.hid_init = hid_init
+        else:
+            self.hid_init = self.add_param(
+                hid_init, (1, self.num_units), name="hid_init",
+                trainable=learn_init, regularizable=False)
+
+    def get_params(self, **tags):
+        # Get all parameters from this layer, the master layer
+        params = super(CustomRecurrentLayer, self).get_params(**tags)
+        # Combine with all parameters from the child layers
+        params += helper.get_all_params(self.input_to_hidden, **tags)
+        params += helper.get_all_params(self.hidden_to_hidden, **tags)
+        return params
+
+    def get_output_shape_for(self, input_shape):
+        return input_shape[0], input_shape[1], self.num_units
+
+    def get_output_for(self, input, mask=None, **kwargs):
+        """
+        Compute this layer's output function given a symbolic input variable.
+
+        Parameters
+        ----------
+        input : theano.TensorType
+            Symbolic input variable.
+        mask : theano.TensorType
+            Theano variable denoting whether each time step in each
+            sequence in the batch is part of the sequence or not.  If ``None``,
+            then it is assumed that all sequences are of the same length.  If
+            not all sequences are of the same length, then it must be
+            supplied as a matrix of shape ``(n_batch, n_time_steps)`` where
+            ``mask[i, j] = 1`` when ``j <= (length of sequence i)`` and
+            ``mask[i, j] = 0`` when ``j > (length of sequence i)``.
```

Comment:
> But, I'm not sure how to do that with the way `lasagne` is currently structured.

It's impossible, but maybe we should change that, or at least figure out if that's something we can change later on without breaking the API. I'll create an Issue.